lib/tasks/migrate.rake

```diff
@@ -1,76 +1,226 @@
+      legacy_metrics: metrics,
+    }
   end
 
-  def create_body(metrics, reported_at, status, logs, host)
+  def build_report(host_id, origin, body, report_keyword_ids)
     {
-      :migrated => "true",
-      :host => host,
-      :reported_at => reported_at,
-      :logs => logs,
-      :keywords => keywords(status),
-      :summary => summary(metrics),
+      host_id: host_id,
+      proxy_id: nil,
+      format: origin,
+      reported_at: body[:reported_at],
+      body: body.to_json,
+      change: body.dig(:summary, :foreman, :change),
+      nochange: body.dig(:summary, :foreman, :nochange),
+      failure: body.dig(:summary, :foreman, :failure),
+      report_keyword_ids: report_keyword_ids,
     }
   end
 
-  def host_report_create(host_id, origin, body)
+  def create_puppet_logs(id, log_object)
+    logs = [["debug", "migration", "Report migrated from legacy report ID=#{id} at #{Time.now.utc}"]]
+    log_object.includes(:message, :source).find_each do |log|
+      logs << [PUPPET_LOG_LEVELS[log.level_id] || 'unknown', log.source.value, log.message.value]
+    end
+    logs
+  end
+
+  def create_ansible_result(msg, level, result = {}, task = {})
     {
-      :host_id => host_id,
-      :proxy_id => nil,
-      :format => origin,
-      :reported_at => body[:reported_at],
-      :body => body.to_json,
-      :change => body[:summary][:foreman][:change],
-      :nochange => body[:summary][:foreman][:nochange],
-      :failure => body[:summary][:foreman][:failure],
+      failed: false,
+      level: level,
+      friendly_message: msg,
+      result: result,
+      task: task,
     }
   end
 
+  def create_ansible_results(id, log_object)
+    results = []
+    results << create_ansible_result("Report migrated from legacy report ID=#{id} at #{Time.now.utc}", "info")
+    log_object.includes(:message, :source).find_each do |log|
+      lvl = PUPPET_LOG_LEVELS[log.level_id] || 'unknown'
+      msg = JSON.parse(log.message.value) rescue log.message.value
+      results << create_ansible_result(log.source.value, lvl, msg)
+    end
+    results
+  end
+
+  desc <<-END_DESC
+  Migrates Foreman Configuration Reports to the new Host Reports format.
+  Does not delete legacy reports, can be iterrupted at any time.
+  Accepts from_date option (older reports will be ignored) and from_id,
+  primary key (ID) to start migration from which can be used to resume
+  previously stopped migration. Example:
+
+  foreman-rake host_reports:migrate from_date=2021-01-01 from_id=1234567
+  END_DESC
   task :migrate => :environment do
-    ConfigReport.all.order(id: :asc).where("id >= ?", (ENV['start'] || '0').to_i).each do |r|
-      if r.origin != "Puppet"
-        next
+    Rails.logger.level = Logger::ERROR
+    Foreman::Logging.logger('permissions').level = Logger::ERROR
+    Foreman::Logging.logger('audit').level = Logger::ERROR
+    Signal.trap("INT") do
+      puts "Graceful shutdown..."
+      reports_migrate_running = false
+    end
```

Comment:
Hmm, apparently something (scl or su most likely) catches the SIGINT and sends SIGTERM to all child processes so this doesn't really work on production setups

```
# foreman-rake host_reports:migrate
API controllers newer than Apipie cache! Run apipie:cache rake task to regenerate cache.
Starting, 550 report(s) left
Processing report 6 out of 550 reports
Processing report 12 out of 550 reports
Processing report 18 out of 550 reports
Processing report 24 out of 550 reports
Processing report 30 out of 550 reports
Processing report 36 out of 550 reports
Processing report 42 out of 550 reports
Processing report 48 out of 550 reports
Processing report 54 out of 550 reports
^C
Session terminated, killing shell...rake aborted!
SignalException: SIGTERM
/opt/theforeman/tfm/root/usr/share/gems/gems/activesupport-6.0.3.7/lib/active_support/json/encoding.rb:58:in `gsub!'
/opt/theforeman/tfm/root/usr/share/gems/gems/activesupport-6.0.3.7/lib/active_support/json/encoding.rb:58:in `to_json'
/opt/theforeman/tfm/root/usr/share/gems/gems/activesupport-6.0.3.7/lib/active_support/json/encoding.rb:106:in `stringify'
/opt/theforeman/tfm/root/usr/share/gems/gems/activesupport-6.0.3.7/lib/active_support/json/encoding.rb:35:in `encode'
/opt/theforeman/tfm/root/usr/share/gems/gems/activesupport-6.0.3.7/lib/active_support/json/encoding.rb:22:in `encode'
/opt/theforeman/tfm/root/usr/share/gems/gems/activesupport-6.0.3.7/lib/active_support/core_ext/object/json.rb:42:in `to_json'
/opt/theforeman/tfm/root/usr/share/gems/gems/foreman_host_reports-0.0.4/lib/tasks/migrate.rake:97:in `build_report'
/opt/theforeman/tfm/root/usr/share/gems/gems/foreman_host_reports-0.0.4/lib/tasks/migrate.rake:186:in `block (5 levels) in <top (required)>'
/usr/share/foreman/app/models/concerns/foreman/thread_session.rb:108:in `as'
/usr/share/foreman/app/models/concerns/foreman/thread_session.rb:114:in `as_anonymous_admin'
/opt/theforeman/tfm/root/usr/share/gems/gems/foreman_host_reports-0.0.4/lib/tasks/migrate.rake:185:in `block (4 levels) in <top (required)>'
/opt/theforeman/tfm/root/usr/share/gems/gems/audited-4.9.0/lib/audited/auditor.rb:376:in `without_auditing'
/opt/theforeman/tfm/root/usr/share/gems/gems/foreman_host_reports-0.0.4/lib/tasks/migrate.rake:184:in `block (3 levels) in <top (required)>'
/opt/theforeman/tfm/root/usr/share/gems/gems/activerecord-6.0.3.7/lib/active_record/relation/batches.rb:70:in `block (2 levels) in find_each'
/opt/theforeman/tfm/root/usr/share/gems/gems/activerecord-6.0.3.7/lib/active_record/relation/batches.rb:70:in `each'
/opt/theforeman/tfm/root/usr/share/gems/gems/activerecord-6.0.3.7/lib/active_record/relation/batches.rb:70:in `block in find_each'
/opt/theforeman/tfm/root/usr/share/gems/gems/activerecord-6.0.3.7/lib/active_record/relation/batches.rb:136:in `block in find_in_batches'
/opt/theforeman/tfm/root/usr/share/gems/gems/activerecord-6.0.3.7/lib/active_record/relation/batches.rb:238:in `block in in_batches'
/opt/theforeman/tfm/root/usr/share/gems/gems/activerecord-6.0.3.7/lib/active_record/relation/batches.rb:222:in `loop'
/opt/theforeman/tfm/root/usr/share/gems/gems/activerecord-6.0.3.7/lib/active_record/relation/batches.rb:222:in `in_batches'
/opt/theforeman/tfm/root/usr/share/gems/gems/activerecord-6.0.3.7/lib/active_record/relation/batches.rb:135:in `find_in_batches'
/opt/theforeman/tfm/root/usr/share/gems/gems/activerecord-6.0.3.7/lib/active_record/relation/batches.rb:69:in `find_each'
/opt/theforeman/tfm/root/usr/share/gems/gems/foreman_host_reports-0.0.4/lib/tasks/migrate.rake:158:in `block (2 levels) in <top (required)>'
/opt/rh/rh-ruby27/root/usr/share/gems/gems/rake-13.0.1/exe/rake:27:in `<top (required)>'
Tasks: TOP => host_reports:migrate
(See full trace by running task with --trace)
 ...killed.
```

Fwiw it works well when the user is already a `foreman` user.
```
# id
uid=995(foreman) gid=993(foreman) groups=993(foreman) context=unconfined_u:unconfined_r:unconfined_t:s0-s0:c0.c1023

# foreman-rake host_reports:migrate
API controllers newer than Apipie cache! Run apipie:cache rake task to regenerate cache.
Starting, 550 report(s) left
Processing report 6 out of 550 reports
Processing report 12 out of 550 reports
Processing report 18 out of 550 reports
Processing report 24 out of 550 reports
Processing report 30 out of 550 reports
Processing report 36 out of 550 reports
^CGraceful shutdown...
Error when processing report ID=42
#<ConfigReport id: 42, host_id: 3, reported_at: [FILTERED], created_at: "2022-03-09 13:43:38", updated_at: "2022-03-09 13:43:38", status: 1151796704221075459, metrics: "---\ntime:\n  config_retrieval: 6.98906397819519\n  t...", type: "ConfigReport", openscap_proxy_id: nil, origin: "Puppet", job_invocation_id: nil>
To resume the process:

***

foreman-rake host_reports:migrate from_id=42 from_date=1980-01-15

***

rake aborted!
Interrupted
/opt/theforeman/tfm/root/usr/share/gems/gems/foreman_host_reports-0.0.4/lib/tasks/migrate.rake:159:in `block (3 levels) in <top (required)>'
/opt/theforeman/tfm/root/usr/share/gems/gems/activerecord-6.0.3.7/lib/active_record/relation/batches.rb:70:in `block (2 levels) in find_each'
/opt/theforeman/tfm/root/usr/share/gems/gems/activerecord-6.0.3.7/lib/active_record/relation/batches.rb:70:in `each'
/opt/theforeman/tfm/root/usr/share/gems/gems/activerecord-6.0.3.7/lib/active_record/relation/batches.rb:70:in `block in find_each'
/opt/theforeman/tfm/root/usr/share/gems/gems/activerecord-6.0.3.7/lib/active_record/relation/batches.rb:136:in `block in find_in_batches'
/opt/theforeman/tfm/root/usr/share/gems/gems/activerecord-6.0.3.7/lib/active_record/relation/batches.rb:238:in `block in in_batches'
/opt/theforeman/tfm/root/usr/share/gems/gems/activerecord-6.0.3.7/lib/active_record/relation/batches.rb:222:in `loop'
/opt/theforeman/tfm/root/usr/share/gems/gems/activerecord-6.0.3.7/lib/active_record/relation/batches.rb:222:in `in_batches'
/opt/theforeman/tfm/root/usr/share/gems/gems/activerecord-6.0.3.7/lib/active_record/relation/batches.rb:135:in `find_in_batches'
/opt/theforeman/tfm/root/usr/share/gems/gems/activerecord-6.0.3.7/lib/active_record/relation/batches.rb:69:in `find_each'
/opt/theforeman/tfm/root/usr/share/gems/gems/foreman_host_reports-0.0.4/lib/tasks/migrate.rake:158:in `block (2 levels) in <top (required)>'
/opt/rh/rh-ruby27/root/usr/share/gems/gems/rake-13.0.1/exe/rake:27:in `<top (required)>'
Tasks: TOP => host_reports:migrate
(See full trace by running task with --trace)
```

Reply:
So are you saying that if I also catch SIGTERM it will work? That is weird.